pkg/operator/configobservation/etcdendpoints/observe_etcd_endpoints.go

```diff
@@ -0,0 +1,109 @@
+package etcdendpoints
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"strings"
+
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/openshift/cluster-openshift-apiserver-operator/pkg/operator/configobservation"
+	endpointsobserver "github.com/openshift/cluster-openshift-apiserver-operator/pkg/operator/configobservation/etcdobserver"
+)
+
+const (
+	etcdEndpointNamespace = "openshift-etcd"
+	etcdEndpointName      = "etcd-endpoints"
+)
+
+// BackwardsCompatibleObserveStorageURLs will prefer to observe the new
+// etcd-endpoints configmap, and will fall back to reading the endpoint
+// resource only if the configmap doesn't yet exist.
+func BackwardsCompatibleObserveStorageURLs(genericListers configobserver.Listers, recorder events.Recorder, currentConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(configobservation.Listers)
+	if _, err := listers.ConfigmapLister.ConfigMaps(etcdEndpointNamespace).Get(etcdEndpointName); err == nil {
+		return ObserveStorageURLs(genericListers, recorder, currentConfig)
+	} else if !errors.IsNotFound(err) {
+		return currentConfig, []error{err}
+	}
+	return endpointsobserver.ObserveStorageURLs(genericListers, recorder, currentConfig)
+}
+
+// ObserveStorageURLs observes the storage config URLs. If there is a problem observing the current storage config URLs,
+// then the previously observed storage config URLs will be re-used.
+func ObserveStorageURLs(genericListers configobserver.Listers, recorder events.Recorder, currentConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(configobservation.Listers)
+	storageConfigURLsPath := []string{"storageConfig", "urls"}
+	var errs []error
+
+	previouslyObservedConfig := map[string]interface{}{}
+	currentEtcdURLs, found, err := unstructured.NestedStringSlice(currentConfig, storageConfigURLsPath...)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	if found {
+		if err := unstructured.SetNestedStringSlice(previouslyObservedConfig, currentEtcdURLs, storageConfigURLsPath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	var etcdURLs []string
+	etcdEndpoints, err := listers.ConfigmapLister.ConfigMaps(etcdEndpointNamespace).Get(etcdEndpointName)
+	if errors.IsNotFound(err) {
+		recorder.Warningf("ObserveStorageFailed", "Required %s/%s configmap not found", etcdEndpointNamespace, etcdEndpointName)
+		return previouslyObservedConfig, append(errs, fmt.Errorf("configmaps/%s.%s: not found", etcdEndpointName, etcdEndpointNamespace))
```

Comment:
```suggestion
		recorder.Warningf("ObserveStorageFailed", "Required %s/%s configmap not found; falling back to deprecated endpoint.", etcdEndpointNamespace, etcdEndpointName)
		errs = append(errs, fmt.Errorf("configmaps/%s.%s: not found", etcdEndpointName, etcdEndpointNamespace))
		observedConfig, observerErrs := endpointsobserver.ObserveStorageURLs(genericListers, recorder, currentConfig)
		return observerConfig, append(errs, observerErrs...)
```
I still think it would be best to call the "legacy" observer func here instead of using `BackwardsCompatibleObserveStorageURLs`.  The err!=nil when looking for configmap case below returns slightly different config & errs than in `BackwardsCompatibleObserveStorageURLs`.

Reply:
>I still think it would be best to call the "legacy" observer func here instead of using BackwardsCompatibleObserveStorageURLs. The err!=nil when looking for configmap case below returns slightly different config & errs than in BackwardsCompatibleObserveStorageURLs.

The advantage of using `BackwardsCompatibleObserveStorageURLs` is that the actual observation code for both old and new remain orthogonal. If we make the decision in ObserveStorageURLs, now the old and new paths are coupled, and we'll even need to expand the test coverage of the new one to incorporate awareness of the old one.

However, I see your concern about the variation in return semantics in the configmap missing case. To address both our concerns: what if we changed `BackwardsCompatibleObserveStorageURLs` so that when the configmap lookup fails for a non-missing error, we do the same `previouslyObservedConfig` logic the observer functions perform so that the output is consistent across all three methods (adapter, old and new observers)?

Reply:
@sanchezl

Reply:
Refactored to incorporate feedback